### PR TITLE
Add create and take offer domains

### PR DESCRIPTION
**Pull request by HenrikJannsen**

### CODE REVIEW

**Comment by rodvar** [comment]
rebased to main

**Comment by rodvar** [comment]
Great effort and a very important PR for the project. I'm focus on testing first before code review, leaving here comments as I test the branch locally.

## Node

 - Add offer worked great, I can see the new offer in both, my node app and a desktop app connected to the same local network. 
 - take offer have some memory issues but couldn't repro all the time , see below for stacktrace (probably not using the right coroutine scope?)
 - when I could take the offers, it looks like not all the comms are fully implemented cause I cannot do the complete real interaction progression in both ends, and if I ignore that and just continue (which shouldn't be possible) the trade doesn't really gets completed and I can take the offer again.
 
 
<img width="415" alt="Screenshot 2024-12-26 at 11 51 40" src="https://github.com/user-attachments/assets/2391d34d-35a3-439d-b7aa-839d3db57813" />

 
 ```
 ANR in network.bisq.mobile.node (network.bisq.mobile.node/network.bisq.mobile.android.node.MainActivity)
                                                                                                     PID: 1151
                                                                                                     Reason: Input dispatching timed out (c73fd66 network.bisq.mobile.node/network.bisq.mobile.android.node.MainActivity (server) is not responding. Waited 5002ms for MotionEvent)
                                                                                                     Parent: network.bisq.mobile.node/network.bisq.mobile.android.node.MainActivity
                                                                                                     ErrorId: 665296b0-400e-4530-872e-b1ccd1ae849d
                                                                                                     Frozen: false
                                                                                                     Load: 1.01 / 0.81 / 0.41
                                                                                                     ----- Output from /proc/pressure/memory -----
                                                                                                     some avg10=0.00 avg60=0.13 avg300=0.21 total=19589875
                                                                                                     full avg10=0.00 avg60=0.08 avg300=0.11 total=9773873
                                                                                                     ----- End output from /proc/pressure/memory -----
                                                                                                     ----- Output from /proc/pressure/cpu -----
                                                                                                     some avg10=0.39 avg60=2.63 avg300=2.56 total=1331295221
                                                                                                     full avg10=0.00 avg60=0.00 avg300=0.00 total=0
                                                                                                     ----- End output from /proc/pressure/cpu -----
                                                                                                     ----- Output from /proc/pressure/io -----
                                                                                                     some avg10=0.00 avg60=0.18 avg300=0.35 total=36965399
                                                                                                     full avg10=0.00 avg60=0.09 avg300=0.13 total=17358632
                                                                                                     ----- End output from /proc/pressure/io -----
                                                                                                     
                                                                                                     CPU usage from 153645ms to -1ms ago (2024-12-26 11:48:44.159 to 2024-12-26 11:51:17.805):
                                                                                                       23% 1151/network.bisq.mobile.node: 16% user + 7.4% kernel / faults: 59210 minor 33 major
                                                                                                       10% 368/android.hardware.graphics.composer3-service.ranchu: 0.7% user + 9.6% kernel / faults: 102 minor 7 major
                                                                                                       5.4% 378/surfaceflinger: 2.6% user + 2.7% kernel / faults: 1339 minor 170 major
                                                                                                       1.7% 537/system_server: 1% user + 0.6% kernel / faults: 45075 minor 520 major
                                                                                                       0.5% 756/com.android.systemui: 0.2% user + 0.3% kernel / faults: 8752 minor 129 major
                                                                                                       0.4% 51/kswapd0: 0% user + 0.4% kernel
                                                                                                       0.4% 997/com.google.android.apps.nexuslauncher: 0.1% user + 0.2% kernel / faults: 10947 minor 571 major
                                                                                                       0.3% 1176/com.google.android.gms.persistent: 0.2% user + 0% kernel / faults: 18154 minor 122 major
                                                                                                       0.1% 23150/com.google.android.gms.unstable: 0% user + 0% kernel / faults: 7166 minor 1674 major
                                                                                                       0.1% 360/android.hardware.sensors-service.multihal: 0% user + 0% kernel / faults: 94 minor 32 major
                                                                                                     17% TOTAL: 6.2% user + 6.2% kernel + 0.2% iowait + 4.7% irq + 0.2% softirq
                                                                                                     CPU usage from 10ms to 239ms later (2024-12-26 11:51:17.814 to 2024-12-26 11:51:18.044):
                                                                                                       37% 1151/network.bisq.mobile.node: 28% user + 9.3% kernel / faults: 2860 minor 167 major
                                                                                                         32% 1200/Signal Catcher: 23% user + 9.3% kernel
                                                                                                       32% 537/system_server: 13% user + 18% kernel / faults: 3087 minor 260 major
                                                                                                         18% 543/Signal Catcher: 9.1% user + 9.1% kernel
                                                                                                         9.1% 1764/AnrAuxiliaryTas: 0% user + 9.1% kernel
                                                                                                         4.5% 583/PowerManagerSer: 4.5% user + 0% kernel
                                                                                                     30% TOTAL: 10% user + 7.8% kernel + 7.8% iowait + 4.4% irq
 ```
 
 ## xClients
 
  - couldn't test this one it gets stuck on "Connected to Trusted Node" in this branch (?) I'll try to bypass it later today to test the related to add/take offer stuff

**Comment by rodvar** [comment]
Screenshot 2024-12-26 at 16 00 46

Is it possible that the APIs for registering stopped working?

```
15:59:38.753 ERROR [Network.IO-111] b.h.w.WebSocketConnectionHandler: No service found for handling message: {
    "type": "WebSocketRestApiRequest",
    "requestId": "8f755629-05da-4d7e-98f6-e2ef260fa221",
    "method": "GET",
    "path": "/api/v1/offerbook/markets",
    "body": ""
} 
15:59:38.753 ERROR [Network.IO-112] b.h.w.WebSocketConnectionHandler: No service found for handling message: {
    "type": "SubscriptionRequest",
    "requestId": "6b8d80d3-4bcb-423d-ab5b-17136ef2e65c",
    "topic": "MARKET_PRICE"
} 
15:59:38.960 INFO  [grizzly-nio-kernel(4) SelectorRunner] b.h.w.WebSocketConnectionHandler: Received message {
    "type": "WebSocketRestApiRequest",
    "requestId": "6b3e1ee3-2db6-4c24-bd00-82cd3231618b",
    "method": "GET",
    "path": "/api/v1/user-identities/key-material",
    "body": ""
} 
15:59:38.961 ERROR [Network.IO-112] b.h.w.WebSocketConnectionHandler: No service found for handling message: {
    "type": "WebSocketRestApiRequest",
    "requestId": "6b3e1ee3-2db6-4c24-bd00-82cd3231618b",
    "method": "GET",
    "path": "/api/v1/user-identities/key-material",
    "body": ""
} 

```

**Comment by rodvar** [comment]
I manage to get in the app for `xClients` regenering a new profile but I get a NPE - hope this helps let me know when to test again 🙏 

```
FATAL EXCEPTION: main
                                                                                                     Process: network.bisq.mobile.client, PID: 5799
                                                                                                     java.lang.NullPointerException
                                                                                                     	at network.bisq.mobile.presentation.ui.uicases.offer.create_offer.CreateOfferPresenter.onStartCreateOffer(CreateOfferPresenter.kt:79)
                                                                                                     	at network.bisq.mobile.presentation.ui.uicases.GettingStartedPresenter.navigateToCreateOffer(GettingStartedPresenter.kt:33)
                                                                                                     	at network.bisq.mobile.presentation.ui.uicases.GettingStartedScreenKt$GettingStartedScreen$1.invoke$lambda$5$lambda$4(GettingStartedScreen.kt:95)
                                                                                                     	at network.bisq.mobile.presentation.ui.uicases.GettingStartedScreenKt$GettingStartedScreen$1.$r8$lambda$ur9s1BCyidvwzjkZOsmGijiuuFM(Unknown Source:0)
                                                                                                     	at network.bisq.mobile.presentation.ui.uicases.GettingStartedScreenKt$GettingStartedScreen$1$$ExternalSyntheticLambda0.invoke(D8$$SyntheticClass:0)
                                                                                                     	at network.bisq.mobile.presentation.ui.components.atoms.ButtonKt.BisqButton_lBpx9hA$lambda$1$lambda$0(Button.kt:67)
                                                                                                     	at network.bisq.mobile.presentation.ui.components.atoms.ButtonKt.$r8$lambda$EQh5W3wsOyID0oIZjTrk8TrzqwE(Unknown Source:0)
                                                                                                     	at network.bisq.mobile.presentation.ui.components.atoms.ButtonKt$$ExternalSyntheticLambda0.invoke(D8$$SyntheticClass:0)
                                                                                                     	at androidx.compose.foundation.ClickableNode$clickPointerInput$3.invoke-k-4lQ0M(Clickable.kt:639)
                                                                                                     	at androidx.compose.foundation.ClickableNode$clickPointerInput$3.invoke(Clickable.kt:633)
                                                                                                     	at androidx.compose.foundation.gestures.TapGestureDetectorKt$detectTapAndPress$2$1.invokeSuspend(TapGestureDetector.kt:255)
                                                                                                     	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                     	at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:165)
                                                                                                     	at kotlinx.coroutines.DispatchedTaskKt.dispatch(DispatchedTask.kt:154)
                                                                                                     	at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:470)
                                                                                                     	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl$kotlinx_coroutines_core(CancellableContinuationImpl.kt:504)
                                                                                                     	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl$kotlinx_coroutines_core$default(CancellableContinuationImpl.kt:493)
                                                                                                     	at kotlinx.coroutines.CancellableContinuationImpl.resumeWith(CancellableContinuationImpl.kt:359)
                                                                                                     	at androidx.compose.ui.input.pointer.SuspendingPointerInputModifierNodeImpl$PointerEventHandlerCoroutine.offerPointerEvent(SuspendingPointerInputFilter.kt:719)
                                                                                                     	at androidx.compose.ui.input.pointer.SuspendingPointerInputModifierNodeImpl.dispatchPointerEvent(SuspendingPointerInputFilter.kt:598)
                                                                                                     	at androidx.compose.ui.input.pointer.SuspendingPointerInputModifierNodeImpl.onPointerEvent-H0pRuoY(SuspendingPointerInputFilter.kt:620)
                                                                                                     	at androidx.compose.foundation.AbstractClickableNode.onPointerEvent-H0pRuoY(Clickable.kt:1044)
                                                                                                     	at androidx.compose.ui.input.pointer.Node.dispatchMainEventPass(HitPathTracker.kt:387)
                                                                                                     	at androidx.compose.ui.input.pointer.Node.dispatchMainEventPass(HitPathTracker.kt:373)
                                                                                                     	at androidx.compose.ui.input.pointer.Node.dispatchMainEventPass(HitPathTracker.kt:373)
                                                                                                     	at androidx.compose.ui.input.pointer.Node.dispatchMainEventPass(HitPathTracker.kt:373)
                                                                                                     	at androidx.compose.ui.input.pointer.NodeParent.dispatchMainEventPass(HitPathTracker.kt:229)
                                                                                                     	at androidx.compose.ui.input.pointer.HitPathTracker.dispatchChanges(HitPathTracker.kt:144)
                                                                                                     	at androidx.compose.ui.input.pointer.PointerInputEventProcessor.process-BIzXfog(PointerInputEventProcessor.kt:120)
                                                                                                     	at androidx.compose.ui.platform.AndroidComposeView.sendMotionEvent-8iAsVTc(AndroidComposeView.android.kt:1993)
                                                                                                     	at androidx.compose.ui.platform.AndroidComposeView.handleMotionEvent-8iAsVTc(AndroidComposeView.android.kt:1944)
                                                                                                     	at androidx.compose.ui.platform.AndroidComposeView.dispatchTouchEvent(AndroidComposeView.android.kt:1828)
                                                                                                     	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3120)
                                                                                                     	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2801)
                                                                                                     	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3120)
                                                                                                     	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2801)
2024-12-26 16:05:22.581   5799-5799  AndroidRuntime          network.bisq.mobile.client           E  	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3120)
                                                                                                     	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2801)
                                                                                                     	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3120)
                                                                                                     	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2801)
                                                                                                     	at com.android.internal.policy.DecorView.superDispatchTouchEvent(DecorView.java:490)
                                                                                                     	at com.android.internal.policy.PhoneWindow.superDispatchTouchEvent(PhoneWindow.java:1904)
                                                                                                     	at android.app.Activity.dispatchTouchEvent(Activity.java:4377)
                                                                                                     	at com.android.internal.policy.DecorView.dispatchTouchEvent(DecorView.java:448)
                                                                                                     	at android.view.View.dispatchPointerEvent(View.java:15919)
                                                                                                     	at android.view.ViewRootImpl$ViewPostImeInputStage.processPointerEvent(ViewRootImpl.java:7021)
                                                                                                     	at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:6815)
                                                                                                     	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:6229)
                                                                                                     	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:6286)
                                                                                                     	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:6252)
                                                                                                     	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:6417)
                                                                                                     	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:6260)
                                                                                                     	at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:6474)
                                                                                                     	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:6233)
                                                                                                     	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:6286)
                                                                                                     	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:6252)
                                                                                                     	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:6260)
                                                                                                     	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:6233)
                                                                                                     	at android.view.ViewRootImpl.deliverInputEvent(ViewRootImpl.java:9211)
                                                                                                     	at android.view.ViewRootImpl.doProcessInputEvents(ViewRootImpl.java:9162)
                                                                                                     	at android.view.ViewRootImpl.enqueueInputEvent(ViewRootImpl.java:9131)
                                                                                                     	at android.view.ViewRootImpl$WindowInputEventReceiver.onInputEvent(ViewRootImpl.java:9337)
                                                                                                     	at android.view.InputEventReceiver.dispatchInputEvent(InputEventReceiver.java:267)
                                                                                                     	at android.os.MessageQueue.nativePollOnce(Native Method)
                                                                                                     	at android.os.MessageQueue.next(MessageQueue.java:335)
                                                                                                     	at android.os.Looper.loopOnce(Looper.java:162)
                                                                                                     	at android.os.Looper.loop(Looper.java:294)
                                                                                                     	at android.app.ActivityThread.main(ActivityThread.java:8177)
                                                                                                     	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                     	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
                                                                                                     	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
                                                                                                     	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [androidx.compose.ui.platform.MotionDurationScaleImpl@f6e364c, androidx.compose.runtime.BroadcastFrameClock@1407695, StandaloneCoroutine{Cancelling}@8d213aa, AndroidUiDispatcher@2ec5b9b]
```

**Comment by HenrikJannsen** [comment]
>     * when I could take the offers, it looks like not all the comms are fully implemented cause I cannot do the complete real interaction progression in both ends, and if I ignore that and just continue (which shouldn't be possible) the trade doesn't really gets completed and I can take the offer again.

You mean after take offer, the trade process? That is not implemented yet. That will be next step. It only triggers the take offer so on the desktop you see a new trade but then nothing more is working yet.

> couldn't test this one it gets stuck on "Connected to Trusted Node" in this branch (?) I'll try to bypass it later today to test the related to add/take offer stuff

You have the latest main from Bisq 2? Clean all data directories on both backend and mobile, to not get issues with corrupted data.

**Comment by HenrikJannsen** [comment]
> onStartCreateOffer

Hmm... I just tested with a fresh data dir on the httpapi side and all works. 

the NP is likely from `marketPriceServiceFacade.selectedMarketPriceItem.value` which migh be caused when the initial request for market price dont work. can you put a break point into the market price request to see if that returns correctly?

I guess in your setup there is somethings wrong. Be sure to have lastest main, republished all, deleted all data dir on httpapi and mobile. If still issues try to check with breakpoints at the first requests and see if any fails, and/or add breakpoints on the backend side.

**Comment by rodvar** [comment]
I'll did get the latest main and republish, will try deleting everything and test again

**Comment by rodvar** [comment]
Cherry-picked buddha's pr navigation commit to make testing easier -> https://github.com/bisq-network/bisq-mobile/pull/142/commits/2f960379629170e8ed4e855d4afc1d73bb6e3574

**Comment by rodvar** [comment]
@HenrikJannsen ignore my profile related comments above, it was because of navigation issues which are fixed by the cherry picked I've done :)

**Comment by rodvar** [comment]
ok, now that I understand what the PR involves and solved the env issues and navigation issue - it's working great.

Just a note for further PRs, always add on the description a bare min help for the others know what was done please 🙏 
Specially this one being a large code change review 🙏 

I'm gonna do a quick code review now but leaning towards merging as is since its working and more iterative refine PRs are coming as explained above.

File: shared/domain/src/commonMain/kotlin/network/bisq/mobile/domain/replicated/security/pow/ProofOfWorkVO.kt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package network.bisq.mobile.domain.replicated.security.pow
+
+import kotlinx.serialization.Serializable
+import network.bisq.mobile.domain.utils.hexToByteArray
+
+@Serializable
+data class ProofOfWorkVO(
```
Comment (rodvar) [inline shared/domain/src/commonMain/kotlin/network/bisq/mobile/domain/replicated/security/pow/ProofOfWorkVO.kt:7]: nit - VO here stands for Value Object right? Wouldn't DTO be more appropriate?

**Review by rodvar — APPROVED** [review]
ACK - merging in for productivity purposes, more PRs are coming to complete the integration

File: shared/domain/src/commonMain/kotlin/network/bisq/mobile/domain/replicated/security/pow/ProofOfWorkVO.kt (hunk 1 above)
Comment (HenrikJannsen) [reply shared/domain/src/commonMain/kotlin/network/bisq/mobile/domain/replicated/security/pow/ProofOfWorkVO.kt:7]: I had earlier Dto, but as the models are used as domain models VO is more appropriate as DTO is more for data transfer. On the Bisq 2 side they are called DTO as there their only purpose is to serve for data transfer.

File: androidNode/src/androidMain/kotlin/network/bisq/mobile/android/node/service/user_profile/NodeUserProfileServiceFacade.kt [hunk 2]
```diff
@@ -98,42 +96,16 @@ class NodeUserProfileServiceFacade(private val applicationService: AndroidApplic
...
-                it.userName,
-                it.pubKeyHash.toString(),
-                it.publishDate
-            )
-        }
+    override suspend fun getSelectedUserProfile(): UserProfileVO? {
+        return userService.userIdentityService.selectedUserIdentity?.userProfile?.let { Mappings.UserProfileMapping.from(it) }
     }
 
+    // Private
```
[the remark this answers is not in the document]
Comment (HenrikJannsen) [reply androidNode/src/androidMain/kotlin/network/bisq/mobile/android/node/service/user_profile/NodeUserProfileServiceFacade.kt:108]: Will clean up in follow-ups

**Pull request merged**